### PR TITLE
feat: allow additional props from IconType to expose additional props in base object

### LINKED ISSUE #4639: DataTable: The parameter of the function passed to `sortIcon` has the wrong type
### Describe the bug

When using a function for the `sortIcon` prop of a `<DataTable>`, the function's first parameter, `options`, is of type `IconOptions<DataTable<TValue>>`. This type knows about icons, but not about sorting.

Logging the `options` parameter to console shows it has more props than its type claims. The most important one is `sortOrder`, telling which direction a column is sorted. TS does not know about this, so `options.sortOrder` defaults to type `any`. It should be `0 | 1 | -1`

In addition to `sortOrder`, there is also a property `sorted` which the type does not know about. Its actual type appears to for some reason be `null | true`, instead of `boolean`.

### Reproducer

https://stackblitz.com/edit/vitejs-vite-s3jiwf?file=src%2FApp.tsx

### PrimeReact version

9.6.0

### React version

18.x

### Language

TypeScript

### Build / Runtime

Create React App (CRA)

### Browser(s)

_No response_

### Steps to reproduce the behavior

Hover `options.sortOrder` and `options.sorted` in the reproducer. They both show type `any`.

Check the console, where both props are logged. They clearly have values.

### Expected behavior

The type should know about `options.sortOrder`, which should have type  `0 | 1 | -1`.

In addition, the type should know about `options.sorted` , which for some reason should have type `null | true`

### PR DESCRIPTION
**Pull request by EmirBoyaci**
Closes #4639

### CODE REVIEW

**Comment by EmirBoyaci** [comment]
Not sure about naming but didn't wanted to give more time to choose what is the best naming for generics, feel free to suggest re-naming.

**Comment by EmirBoyaci** [comment]
> so far I like the look of it. Let me test it out later today!

Sounds good. With adding `AdditionalProps` generic, we also add ability to define custom additional properties for specific cases as like in this.
So, in the future I think we can delete `[key: string]: any;` to make sure all properties exactly matches with the type definitions.

**Comment by melloware** [comment]
You might be right i hopefully will be able to play around with your PR later today to see if I run into any issues.

File: components/lib/datatable/datatable.d.ts [hunk 1]
```diff
@@ -1269,11 +1271,11 @@ export interface DataTableProps<TValue extends DataTableValueArray> extends Omit
     /**
      * Icon to display the current sorting status.
      */
-    sortIcon?: IconType<DataTable<TValue>> | undefined;
+    sortIcon?: IconType<DataTable<TValue>, { sortOrder: Omit<SortOrder, 'null' | 'undefined'>; sorted: boolean }> | undefined;
```
Comment (melloware) [inline components/lib/datatable/datatable.d.ts:1274]: Are we sure it shouldn't be: `{ sortOrder?: SortOrder; sorted?: boolean }` as optional or have you confirmed they ALWAYS have values?

Comment (EmirBoyaci) [reply components/lib/datatable/datatable.d.ts:1274]: I tried with 3-4 example docs via running server on my local machine and It never returned undefined. But, I am not sure maybe we have some edge case. If you wish, we can defined as `?` to make sure we don't break it.

Comment (melloware) [reply components/lib/datatable/datatable.d.ts:1274]: Yeah lets do that to be safe?

Comment (EmirBoyaci) [reply components/lib/datatable/datatable.d.ts]: Done

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
I just made one small change and some docs but this is working great!

**Pull request merged**